src/quota_aggregator_impl_test.cc

```diff
@@ -0,0 +1,354 @@
+  }
+}
+service_config_id: "2017-02-08r5"
+)";
+
+const char kErrorResponse2[] = R"(
+operation_id: "a197c6f2-aecc-4a31-9744-b1d5aea4e4b4"
+allocate_errors {
+  code: RESOURCE_EXHAUSTED
+  subject: "user:integration_test_user"
+}
+)";
+
+const char kEmptyResponse[] = R"(
+)";
+
+}  // namespace
+
+class QuotaAggregatorImplTest : public ::testing::Test {
+ public:
+  void SetUp() {
+    ASSERT_TRUE(TextFormat::ParseFromString(kRequest1, &request1_));
+    ASSERT_TRUE(TextFormat::ParseFromString(kRequest2, &request2_));
+    ASSERT_TRUE(TextFormat::ParseFromString(kRequest3, &request3_));
+
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kSuccessResponse1, &pass_response1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kSuccessResponse2, &pass_response2_));
+
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kErrorResponse1, &error_response1_));
+    ASSERT_TRUE(
+        TextFormat::ParseFromString(kErrorResponse2, &error_response2_));
+
+    ASSERT_TRUE(TextFormat::ParseFromString(kEmptyResponse, &empty_response_));
+
+    QuotaAggregationOptions options(10, kFlushIntervalMs);
+
+    aggregator_ =
+        CreateAllocateQuotaAggregator(kServiceName, kServiceConfigId, options);
+    ASSERT_TRUE((bool)(aggregator_));
+
+    aggregator_->SetFlushCallback(std::bind(
+        &QuotaAggregatorImplTest::FlushCallback, this, std::placeholders::_1));
+  }
+
+  void FlushCallback(const AllocateQuotaRequest& request) {
+    flushed_.push_back(request);
+  }
+
+  void FlushCallbackCallingBackToAggregator(
+      const AllocateQuotaRequest& request) {
+    flushed_.push_back(request);
+    aggregator_->CacheResponse(request, pass_response1_);
+  }
+
+  AllocateQuotaRequest request1_;
+  AllocateQuotaRequest request2_;
+  AllocateQuotaRequest request3_;
+
+  AllocateQuotaResponse pass_response1_;
+  AllocateQuotaResponse pass_response2_;
+
+  AllocateQuotaResponse error_response1_;
+  AllocateQuotaResponse error_response2_;
+
+  AllocateQuotaResponse empty_response_;
+
+  std::unique_ptr<QuotaAggregator> aggregator_;
+  std::vector<AllocateQuotaRequest> flushed_;
+};
+
+TEST_F(QuotaAggregatorImplTest, TestFirstReqeustNotFound) {
+  AllocateQuotaResponse response;
+
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request1_, &response));
+}
+
+TEST_F(QuotaAggregatorImplTest, TestSecondRequestFound) {
+  AllocateQuotaResponse response;
+
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request1_, &response));
+  EXPECT_OK(aggregator_->Quota(request1_, &response));
+  EXPECT_TRUE(MessageDifferencer::Equals(response, empty_response_));
+}
+
+TEST_F(QuotaAggregatorImplTest, TestFirstReqeustSucceeded) {
+  AllocateQuotaResponse response;
+
+  EXPECT_ERROR_CODE(Code::NOT_FOUND, aggregator_->Quota(request1_, &response));
+  EXPECT_OK(aggregator_->CacheResponse(request1_, pass_response1_));
+  EXPECT_OK(aggregator_->Quota(request1_, &response));
+  EXPECT_TRUE(MessageDifferencer::Equals(response, pass_response1_));
```

Comment:
cached positive response is OK not the same as the response passed into CacheRespone()